File: jrb_sensors/jrb_sensors/sample_detector.py
#!/usr/bin/env python3

import math
from re import I
import traceback
import rclpy
from rclpy.node import Node

from jrb_msgs.msg import SampleDetected, SampleDetectedArray
from ament_index_python import get_package_share_directory
from visualization_msgs.msg import Marker, MarkerArray
from tf2_ros import TransformBroadcaster, StaticTransformBroadcaster
from geometry_msgs.msg import TransformStamped, Pose, Quaternion
from builtin_interfaces.msg import Duration
from sensor_msgs.msg import Image, CompressedImage, CameraInfo
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from tf_transformations import (
    concatenate_matrices,
    translation_matrix,
    quaternion_matrix,
    translation_from_matrix,
    euler_from_matrix,
    quaternion_from_euler,
)
from copy import copy
from math import radians
from rcl_interfaces.msg import ParameterDescriptor, ParameterType, FloatingPointRange
from rcl_interfaces.msg import SetParametersResult


import numpy as np
import cv2
import cv2.aruco as aruco
import sys
import os
from queue import Queue
import threading
from cv_bridge import CvBridge
from ament_index_python import get_package_share_directory


DATA_PATH = get_package_share_directory("jrb_sensors")

def make_marker_msg(id_, stamp, pose, color):
    marker = Marker()

    marker.header.frame_id = "camera_link_optical"
    marker.header.stamp = stamp

    marker.lifetime = Duration(nanosec=int(2 * 1e9))

    marker.type = Marker.MESH_RESOURCE
    marker.action = Marker.ADD
    marker.id = id_

    marker.scale.x = 1.0
    marker.scale.y = 1.0
    marker.scale.z = 1.0

    # Set the pose of the marker
    marker.pose = pose

    # Set the color
    if color == SampleDetected.RED:
        marker.mesh_resource = "file://" + os.path.join(
            DATA_PATH, "meshes/echantillon_rouge.dae"
        )
    elif color == SampleDetected.GREEN:
        marker.mesh_resource = "file://" + os.path.join(
            DATA_PATH, "meshes/echantillon_vert.dae"
        )
    elif color == SampleDetected.BLUE:
        marker.mesh_resource = "file://" + os.path.join(
            DATA_PATH, "meshes/echantillon_bleu.dae"
        )
    elif color == SampleDetected.ROCK:
        marker.mesh_resource = "file://" + os.path.join(
            DATA_PATH, "meshes/echantillon_surprise.dae"
        )

    marker.mesh_use_embedded_materials = True

    return marker


def opencv_to_ros(vec):
    return np.array([-vec[1], vec[0], vec[2]])

class SampleDetector(Node):
    def __init__(self):
        super().__init__("sample_detector")
        self.get_logger().info("init")

        self.publisher_ = self.create_publisher(
            SampleDetectedArray, "sample_detected", 10
        )
        self.maker_publisher = self.create_publisher(
            MarkerArray, "debug/sample_detected", 10
        )
        self.image_publisher = self.create_publisher(Image, "debug/sample_detected_image", 10)

        self.image_subscriber = self.create_subscription(
            CompressedImage,
            "/camera/image_raw/compressed",
            self.on_image_compressed,
            10,
        )
        self.camera_info_subscriber = self.create_subscription(
            CameraInfo, "/camera/camera_info", self.on_camera_info, 10
        )

        # Tf publisher
        self.tf_broadcaster = TransformBroadcaster(self)
        self.tf_static_broadcaster = StaticTransformBroadcaster(self)

        # Tf subscriber
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        self.cv_bridge = CvBridge()
        self.aruco_dict = aruco.Dictionary_get(aruco.DICT_4X4_1000)
        self.cameraMatrix = None
        self.distCoeffs = None
        self.arucoParameters = aruco.DetectorParameters_create()

        # calibration en trichant sur la taille des tags :
        self.big_marker_size = 0.030  # en m
        self.small_marker_size = 0.015  # en m

        self.length_of_axis = 0.02  # en m

        # Wait for static transforms
        self.get_logger().info("Wait for static transforms")
        now = self.get_clock().now().to_msg()

        f = self.tf_buffer.wait_for_transform_async("chassis", "camera_link", now)
        rclpy.spin_until_future_complete(self, f)

        self.get_logger().info("Static transforms found")

        self.add_on_set_parameters_callback(self.on_update_parameters)

    def on_update_parameters(self, params):
        self.get_logger().info("Params updated")

        return SetParametersResult(successful=True)

    def lookupTransform(
        self, target_frame, source_frame, time=rclpy.time.Time().to_msg()
    ):
        transform_msg = self.tf_buffer.lookup_transform(
            target_frame, source_frame, time
        )
        trans = np.array(
            [
                transform_msg.transform.translation.x,
                transform_msg.transform.translation.y,
                transform_msg.transform.translation.z,
            ]
        )
        rot = np.array(
            [
                transform_msg.transform.rotation.x,
                transform_msg.transform.rotation.y,
                transform_msg.transform.rotation.z,
                transform_msg.transform.rotation.w,
            ]
        )
        transform = concatenate_matrices(
            translation_matrix(trans), quaternion_matrix(rot)
        )

        return transform

    def on_camera_info(self, msg):
        self.get_logger().info("Got camera info, unsubscribing")
        self.cameraMatrix = np.array(msg.k).reshape(3, 3)
        self.distCoeffs = np.array(msg.d).reshape(1, 5)
        self.destroy_subscription(self.camera_info_subscriber)

    def on_image_compressed(self, msg):
        if self.cameraMatrix is None or self.distCoeffs is None:
            self.get_logger().warn("No CameraInfo yet, discard frame")
            return

        now = self.get_clock().now().to_msg()

        frame = self.cv_bridge.compressed_imgmsg_to_cv2(msg)
        #frame = cv2.imread("/home/jeroboam/ros2_ws/src/jeroboam_ros/jrb_sensors/resource/image.png")

        # img_msg = self.cv_bridge.cv2_to_imgmsg(frame)
        # img_msg.header.frame_id = "camera_link"
        # self.image_publisher.publish(img_msg)

        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        corners, ids, rejectedImgPoints = aruco.detectMarkers(
            gray, self.aruco_dict, parameters=self.arucoParameters
        )

        if (
            type(ids).__module__ == np.__name__
        ):  # on vérifie que ids est un numpy array. C'est le cas uniquement si il y a au moins en tag de détecté.
            rvecs, tvecs, _objPoints = aruco.estimatePoseSingleMarkers(
                corners, self.big_marker_size, self.cameraMatrix, self.distCoeffs
            )

            # Draw origin
            frame = aruco.drawAxis(
                frame,
                self.cameraMatrix,
                self.distCoeffs,
                np.array([[0.0, 0.0, 0.0]]),
                np.array([[0.0, 0.0, 0.0]]),
                self.length_of_axis,
            )

            frame = aruco.drawDetectedMarkers(frame, corners, ids)

            msg = SampleDetectedArray()
            msg.header.stamp = now
            msg.header.frame_id = "camera_link_optical"

            markers_msg = MarkerArray()
            

            for i in range(len(tvecs)):
                sample_msg = SampleDetected()

                frame = aruco.drawAxis(
                    frame,
                    self.cameraMatrix,
                    self.distCoeffs,
                    rvecs[i],
                    tvecs[i],
                    self.length_of_axis,
                )

                if ids[i] == 47:
                    sample_msg.type = SampleDetected.PINK
                elif ids[i] == 13:
                    sample_msg.type = SampleDetected.YELLOW
                elif ids[i] == 36:
                    sample_msg.type = SampleDetected.BROWN
                elif ids[i] == 91:
                    continue
                elif ids[i] == 92:
                    continue
                else:
                    continue

                xyz = tvecs[i][0]

                sample_msg.pose.position.x = xyz[0]
                sample_msg.pose.position.y = xyz[1]
                #sample_msg.pose.position.z = 0.355
                sample_msg.pose.position.z = xyz[2]


                rotation_matrix = np.identity(4)
                rotation_matrix[:3, :3], _ = cv2.Rodrigues(np.array([rvecs[i][0][:3]]))
                _, _, yaw = euler_from_matrix(rotation_matrix)
                q = quaternion_from_euler(0, 0, yaw)
                sample_msg.pose.orientation.x = q[0]
                sample_msg.pose.orientation.y = q[1]
                sample_msg.pose.orientation.z = q[2]
                sample_msg.pose.orientation.w = q[3]

                # Translate pose from tag to disque center
                new_pose=translate_pose_along_local_y(sample_msg.pose,-0.033) #décallage de  33mm
                sample_msg.pose=new_pose

                # Draw the translated marker for debugging
                xyz[0]=new_pose.position.x
                xyz[1]=new_pose.position.y
                xyz[2]=new_pose.position.z
                tvecs[i][0]=xyz
                frame = aruco.drawAxis(
                    frame,
                    self.cameraMatrix,
                    self.distCoeffs,
                    rvecs[i],
                    tvecs[i],
                    self.length_of_axis,
                )

                msg.samples.append(sample_msg)
                marker_msg = make_marker_msg(
                    len(markers_msg.markers),
                    now,
                    sample_msg.pose,
                    sample_msg.type,
                )
                markers_msg.markers.append(marker_msg)

            self.publisher_.publish(msg)
            self.maker_publisher.publish(markers_msg)

        img_msg = self.cv_bridge.cv2_to_imgmsg(frame,"bgr8")
        img_msg.header.frame_id = "camera_link_optical"
        self.image_publisher.publish(img_msg)


def translate_pose_along_local_y(pose, distance):
    # Extract position and orientation from the original pose
    position = [pose.position.x, pose.position.y, pose.position.z]
    orientation = [pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w]

    # Compute the rotation matrix based on the orientation
    rotation_matrix = quaternion_to_rotation_matrix(orientation)

    # Compute the translation vector along the local X-axis
    translation_vector = [0.0, distance, 0.0]
    translated_position = translate_vector(position, rotation_matrix, translation_vector)

    # Create the translated pose
    translated_pose = Pose()
    translated_pose.position.x = translated_position[0]
    translated_pose.position.y = translated_position[1]
    translated_pose.position.z = translated_position[2]
    translated_pose.orientation = pose.orientation

    return translated_pose

def quaternion_to_rotation_matrix(quaternion):
    q_x, q_y, q_z, q_w = quaternion
    rotation_matrix = [
        [1 - 2 * (q_y ** 2) - 2 * (q_z ** 2), 2 * q_x * q_y - 2 * q_z * q_w, 2 * q_x * q_z + 2 * q_y * q_w],
        [2 * q_x * q_y + 2 * q_z * q_w, 1 - 2 * (q_x ** 2) - 2 * (q_z ** 2), 2 * q_y * q_z - 2 * q_x * q_w],
        [2 * q_x * q_z - 2 * q_y * q_w, 2 * q_y * q_z + 2 * q_x * q_w, 1 - 2 * (q_x ** 2) - 2 * (q_y ** 2)]
    ]
    return rotation_matrix

def translate_vector(vector, rotation_matrix, translation_vector):
    translated_vector = [0.0, 0.0, 0.0]
    for i in range(3):
        for j in range(3):
            translated_vector[i] += rotation_matrix[i][j] * translation_vector[j]
        translated_vector[i] += vector[i]
    return translated_vector


def main(args=None):
    rclpy.init(args=args)

    node = SampleDetector()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Node stopped cleanly")
    except Exception:
        print("Error while stopping the node")
        print(traceback.format_exc())
        raise
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: jrb_sensors/src/sample_detector.cpp
#include <cmath>
#include <memory>
#include <stdexcept>
#include <string>

#include <opencv2/core.hpp>
#include <opencv2/aruco.hpp>
#include <opencv2/calib3d.hpp>

#include "ament_index_cpp/get_package_share_directory.hpp"
#include "cv_bridge/cv_bridge.h"
#include "image_transport/image_transport.hpp"
#include "geometry_msgs/msg/pose.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "tf2/LinearMath/Matrix3x3.h"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2/LinearMath/Transform.h"
#include "tf2_ros/buffer.h"
#include "tf2_ros/static_transform_broadcaster.h"
#include "tf2_ros/transform_broadcaster.h"
#include "tf2_ros/transform_listener.h"
#include "visualization_msgs/msg/marker.hpp"
#include "visualization_msgs/msg/marker_array.hpp"

#include "jrb_msgs/msg/sample_detected.hpp"
#include "jrb_msgs/msg/sample_detected_array.hpp"

const auto DATA_PATH = ament_index_cpp::get_package_share_directory("jrb_sensors");

class SampleDetector : public rclcpp::Node
{
public:
    SampleDetector(rclcpp::NodeOptions const &options) : Node("sample_detector", options)
    {
        this->declare_parameter("big_marker_size", 0.050);
        this->declare_parameter("small_marker_size", 0.015);
        this->declare_parameter("length_of_axis", 0.02);
        this->declare_parameter("publish_debug", true);

        big_marker_size_ = this->get_parameter("big_marker_size").as_double();
        small_marker_size_ = this->get_parameter("small_marker_size").as_double();
        length_of_axis_ = this->get_parameter("length_of_axis").as_double();
        publish_debug_ = this->get_parameter("publish_debug").as_bool();

        auto sensor_qos = rclcpp::QoS(rclcpp::SensorDataQoS());

        publisher_ = this->create_publisher<jrb_msgs::msg::SampleDetectedArray>(
            "sample_detected", 10);
        marker_publisher_ = this->create_publisher<visualization_msgs::msg::MarkerArray>(
            "debug/sample_detected", 10);
        image_publisher_ = image_transport::create_publisher(this, "debug/image", sensor_qos.get_rmw_qos_profile());

        camera_info_subscriber_ = create_subscription<sensor_msgs::msg::CameraInfo>("/camera/camera_info", 10, std::bind(&SampleDetector::on_camera_info, this, std::placeholders::_1));

        if (options.use_intra_process_comms())
        {
            image_subscriber_ = create_subscription<sensor_msgs::msg::Image>("/camera/image_raw", 10, std::bind(&SampleDetector::on_image_raw, this, std::placeholders::_1));
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "no intra coms");
            it_image_subscriber_ = image_transport::create_subscription(this, "/camera/image_raw", std::bind(&SampleDetector::on_it_image_raw, this, std::placeholders::_1), "raw", sensor_qos.get_rmw_qos_profile());
        }

        tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        aruco_dict_ = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_1000);
        aruco_params_ = cv::aruco::DetectorParameters::create();
    }

private:
    void on_it_image_raw(const sensor_msgs::msg::Image::ConstSharedPtr &msg)
    {
        if (camera_matrix_.empty() || dist_coeffs_.empty())
        {
            RCLCPP_WARN(this->get_logger(), "No CameraInfo yet, discard frame");
            return;
        }

        cv::Mat frame = cv_bridge::toCvShare(msg, "bgr8")->image;

        cv::Mat gray;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        std::vector<int> ids;
        std::vector<std::vector<cv::Point2f>> corners;
        cv::aruco::detectMarkers(gray, aruco_dict_, corners, ids, aruco_params_);

        if (!ids.empty())
        {
            std::vector<cv::Vec3d> rvecs, tvecs;
            cv::aruco::estimatePoseSingleMarkers(corners, big_marker_size_, camera_matrix_,
                                                 dist_coeffs_, rvecs, tvecs);

            jrb_msgs::msg::SampleDetectedArray msg;
            msg.header.stamp = this->now();
            msg.header.frame_id = "camera_link_optical";

            visualization_msgs::msg::MarkerArray markers_msg;

            for (size_t i = 0; i < ids.size(); ++i)
            {
                jrb_msgs::msg::SampleDetected sample_msg;

                if (ids[i] == 47)
                {
                    sample_msg.type = jrb_msgs::msg::SampleDetected::RED;
                }
                else if (ids[i] == 13)
                {
                    sample_msg.type = jrb_msgs::msg::SampleDetected::BLUE;
                }
                else if (ids[i] == 36)
                {
                    sample_msg.type = jrb_msgs::msg::SampleDetected::GREEN;
                }
                else if (ids[i] == 17)
                {
                    sample_msg.type = jrb_msgs::msg::SampleDetected::ROCK;
                }
                else if (ids[i] == 91 || ids[i] == 92)
                {
                    continue;
                }
                else
                {
                    continue;
                }

                sample_msg.pose.position.x = tvecs[i][0];
                sample_msg.pose.position.y = tvecs[i][1];
                sample_msg.pose.position.z = 0.355;

                cv::Mat rotationMatrix;
                cv::Rodrigues(rvecs[i], rotationMatrix);

                // tf2::Matrix3x3 tf2Matrix(
                //     rotationMatrix.at<double>(0, 0), rotationMatrix.at<double>(0, 1), rotationMatrix.at<double>(0, 2),
                //     rotationMatrix.at<double>(1, 0), rotationMatrix.at<double>(1, 1), rotationMatrix.at<double>(1, 2),
                //     rotationMatrix.at<double>(2, 0), rotationMatrix.at<double>(2, 1), rotationMatrix.at<double>(2, 2));

                // tf2::Quaternion q;
                // tf2Matrix.getRotation(q);

                tf2::Matrix3x3 rotation_matrix;
                double angle_z = atan2(rotationMatrix.at<double>(1, 0), rotationMatrix.at<double>(0, 0));
                rotation_matrix.setEulerYPR(angle_z, 0, 0);
                tf2::Quaternion q;
                rotation_matrix.getRotation(q);

                sample_msg.pose.orientation.x = q.x();
                sample_msg.pose.orientation.y = q.y();
                sample_msg.pose.orientation.z = q.z();
                sample_msg.pose.orientation.w = q.w();

                msg.samples.push_back(sample_msg);

                visualization_msgs::msg::Marker marker_msg = make_marker_msg(
                    markers_msg.markers.size(), this->now(), sample_msg.pose, std::string(sample_msg.type));
                markers_msg.markers.push_back(marker_msg);
            }

            if (publish_debug_)
            {

                cv::aruco::drawAxis(frame, camera_matrix_, dist_coeffs_, cv::Vec3d({0.0, 0.0, 0.0}), cv::Vec3d({0.0, 0.0, 0.0}), length_of_axis_);
                cv::aruco::drawDetectedMarkers(frame, corners, ids);

                sensor_msgs::msg::Image::SharedPtr img_msg =
                    cv_bridge::CvImage(msg.header, "bgr8", frame).toImageMsg();
                img_msg->header.frame_id = "camera_link_optical";
                image_publisher_.publish(img_msg);
            }

            publisher_->publish(msg);
            marker_publisher_->publish(markers_msg);
        }
    }

    void on_image_raw(sensor_msgs::msg::Image::UniquePtr msg)
    {
        const sensor_msgs::msg::Image::ConstSharedPtr msg_shared(msg.get());
        on_it_image_raw(msg_shared);
    }

    void on_camera_info(sensor_msgs::msg::CameraInfo::UniquePtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Got camera info, unsubscribing");
        camera_matrix_ = cv::Mat(3, 3, CV_64F, const_cast<double *>(msg->k.data())).clone();
        dist_coeffs_ = cv::Mat(1, 5, CV_64F, const_cast<double *>(msg->d.data())).clone();
        camera_info_subscriber_.reset();
    }

    visualization_msgs::msg::Marker make_marker_msg(
        size_t id, const rclcpp::Time &stamp,
        const geometry_msgs::msg::Pose &pose,
        std::string color)
    {
        visualization_msgs::msg::Marker marker;

        marker.header.frame_id = "camera_link_optical";
        marker.header.stamp = stamp;

        marker.lifetime = rclcpp::Duration(2, 0);

        marker.type = visualization_msgs::msg::Marker::MESH_RESOURCE;
        marker.action = visualization_msgs::msg::Marker::ADD;
        marker.id = id;

        marker.scale.x = 1.0;
        marker.scale.y = 1.0;
        marker.scale.z = 1.0;

        marker.pose = pose;

        if (color == jrb_msgs::msg::SampleDetected::RED)
        {
            marker.mesh_resource =
                "file://" + DATA_PATH + "/meshes/echantillon_rouge.dae";
        }
        else if (color == jrb_msgs::msg::SampleDetected::GREEN)
        {
            marker.mesh_resource =
                "file://" + DATA_PATH + "/meshes/echantillon_vert.dae";
        }
        else if (color == jrb_msgs::msg::SampleDetected::BLUE)
        {
            marker.mesh_resource =
                "file://" + DATA_PATH + "/meshes/echantillon_bleu.dae";
        }
        else if (color == jrb_msgs::msg::SampleDetected::ROCK)
        {
            marker.mesh_resource =
                "file://" + DATA_PATH + "/meshes/echantillon_surprise.dae";
        }

        marker.mesh_use_embedded_materials = true;

        return marker;
    }
    rclcpp::Publisher<jrb_msgs::msg::SampleDetectedArray>::SharedPtr publisher_;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_publisher_;
    image_transport::Publisher image_publisher_;

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_subscriber_;
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr camera_info_subscriber_;
    image_transport::Subscriber it_image_subscriber_;

    std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

    cv::Ptr<cv::aruco::Dictionary> aruco_dict_;
    cv::Ptr<cv::aruco::DetectorParameters> aruco_params_;
    cv::Mat camera_matrix_;
    cv::Mat dist_coeffs_;

    double big_marker_size_;
    double small_marker_size_;
    double length_of_axis_;
    bool publish_debug_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SampleDetector>(rclcpp::NodeOptions{}));
    rclcpp::shutdown();
    return 0;
}